athena-jdbc/src/main/java/com/amazonaws/connectors/athena/jdbc/saphana/SAPHANAMetadataHandler.java

```diff
@@ -0,0 +1,196 @@
+                if (!resultSet.next()) {
+                    blockWriter.writeRows((Block block, int rowNum) -> {
+                        block.setValue(BLOCK_PARTITION_COLUMN_NAME, rowNum, ALL_PARTITIONS);
+                        LOGGER.info("Adding partition {}", ALL_PARTITIONS);
+                        //we wrote 1 row so we return 1
+                        return 1;
+                    });
+                }
+                else {
+                    do {
+                        final String partitionName = resultSet.getString(PARTITION_COLUMN_NAME);
+
+                        // 1. Returns all partitions of table, we are not supporting constraints push down to filter partitions.
+                        // 2. This API is not paginated, we could use order by and limit clause with offsets here.
+                        blockWriter.writeRows((Block block, int rowNum) -> {
+                            block.setValue(BLOCK_PARTITION_COLUMN_NAME, rowNum, partitionName);
+                            LOGGER.info("Adding partition {}", partitionName);
+                            //we wrote 1 row so we return 1
+                            return 1;
+                        });
+                    }
+                    while (resultSet.next() && queryStatusChecker.isQueryRunning());
+                }
+            }
+        }
+        catch (SQLException sqlException) {
+            throw new RuntimeException(sqlException.getErrorCode() + ": " + sqlException.getMessage(), sqlException);
+        }
+    }
+
+    @Override
+    public GetSplitsResponse doGetSplits(
+            final BlockAllocator blockAllocator, final GetSplitsRequest getSplitsRequest)
+    {
+        LOGGER.info("{}: Catalog {}, table {}", getSplitsRequest.getQueryId(), getSplitsRequest.getTableName().getSchemaName(), getSplitsRequest.getTableName().getTableName());
+        int partitionContd = decodeContinuationToken(getSplitsRequest);
+        Set<Split> splits = new HashSet<>();
+        Block partitions = getSplitsRequest.getPartitions();
+
+        // TODO consider splitting further depending on #rows or data size. Could use Hash key for splitting if no partitions.
+        for (int curPartition = partitionContd; curPartition < partitions.getRowCount(); curPartition++) {
+            FieldReader locationReader = partitions.getFieldReader(BLOCK_PARTITION_COLUMN_NAME);
+            locationReader.setPosition(curPartition);
+
+            SpillLocation spillLocation = makeSpillLocation(getSplitsRequest);
+
+            LOGGER.info("{}: Input partition is {}", getSplitsRequest.getQueryId(), locationReader.readText());
+
+            Split.Builder splitBuilder = Split.newBuilder(spillLocation, makeEncryptionKey())
+                    .add(BLOCK_PARTITION_COLUMN_NAME, String.valueOf(locationReader.readText()));
+
+            splits.add(splitBuilder.build());
+
+            if (splits.size() >= MAX_SPLITS_PER_REQUEST) {
```

Comment:
I think MAX_SPITS_PER_REQUEST should probably be reduced to something like <= 1000

How many splits to we expect to create in a typical query? If each split represents < ~64Mb of data we should chat about if this is the correct model

Reply:
MAX_SPITS_PER_REQUEST has been updated to 1000